learntools/python/ex4.py

```diff
@@ -0,0 +1,129 @@
+                 [1,2,3,4,5]
+                )
+        def sol_fn(x): x[0], x[-1] = x[-1], x[0]
+        for l in lists:
+            copy_for_soln_fn = l.copy()
+            copy_for_user_fn = l.copy()
+            sol_fn(copy_for_soln_fn) # create desired output for comparison
+            user_output = fn(copy_for_user_fn) # also applies swap in this line
+            assert(type(user_output) == type(None)), ("Your function should not return anything."
+                                                      "Instead, change the list without returning it.")
+            assert copy_for_user_fn == copy_for_soln_fn, \
+                "Expected " + copy_for_soln_fn  + " on list " + l + \
+                ".\nGot " + copy_for_user_fn + " instead."
+
+
+
+    _solution = CS(
+"""def purple_shell(racers):
+    # One slick way to do the swap is x[0], x[-1] = x[-1], x[0].
+    temp = racers[0]
+    racers[0] = racers[-1]
+    racers[-1] = temp
+    return""")
+
+class UnderstandLen(Problem):
+    '''TODO: Wrap index errors'''
+    _var = 'understand_len'
+
+    _hint = "Use len to check the lengths of the lists. Call the solution function for an explanation"
+
+    _solution = CS(
+"""[1, 2, 3] - There are three items in this list. Nothing tricky yetself.
+[1, [2, 3]] - The list [2, 3] counts as a single item. It has one item before it. So we have 2 items in the list
+[] - The empty list has 0 items
+[1, 2, 3][1:] - The expression is the same as the list [2, 3], which has length 2.""")
+
+class FashionablyLate(FunctionProblem):
+    _var = 'fashionably_late'
+
+
+
+    _test_cases = [
+            ((['Adela', 'Fleda', 'Owen', 'May', 'Mona', 'Gilbert', 'Ford'], "Adela"), False),
+            ((['Adela', 'Fleda', 'Owen', 'May', 'Mona', 'Gilbert', 'Ford'], "Fleda"), False),
+            ((['Adela', 'Fleda', 'Owen', 'May', 'Mona', 'Gilbert', 'Ford'], "Owen"), False),
+            ((['Adela', 'Fleda', 'Owen', 'May', 'Mona', 'Gilbert', 'Ford'], "May"), False),
+            ((['Adela', 'Fleda', 'Owen', 'May', 'Mona', 'Gilbert', 'Ford'], "Mona"), True),
+            ((['Adela', 'Fleda', 'Owen', 'May', 'Mona', 'Gilbert', 'Ford'], "Gilbert"), True),
+            ((['Adela', 'Fleda', 'Owen', 'May', 'Mona', 'Gilbert', 'Ford'], "Ford"), False),
+            ((["Paul", "John", "Ringo", "George"], "John"), False),
+            ((["Paul", "John", "Ringo", "George"], "Ringo"), False),
+            ((["Lebron", "Kevin"], "Lebron"), False),
+            ((["Lebron", "Kevin"], "Kevin"), False),
+    ]
+
+    _hint = ("Use the index method to find when the person arrived. Check whether "
+            "that is a fashionably late spot given the list length (len). Think about 0-indexing"
+             )
+
+    _solution = CS(
+"""def fashionably_late(arrivals, name):
+    arrival_spot = arrivals.index(name)
+    index_midpoint = len(arrivals) / 2       # could be fraction like 3.5
+    return (arrival_spot != (list_len-1)) and (arrival_spot > index_midpoint)""")
```

Comment:
`NameError: name 'list_len' is not defined` :P

But more serious issue: I believe the `>` should be `>=` (this will recognize Ringo as being fashionably late in the above example)